File: java-adaptor/src/main/java/org/judal/storage/java/MapFieldHelper.java
package org.judal.storage.java;

import java.util.Map;
import java.lang.reflect.InvocationTargetException;

class MapFieldHelper {

	/**
	 * <p>Get value of an HStore field<p>
	 * This method is only supported for PostgreSQL HStore fields
	 * @param sKey JavaRecord Record instance
	 * @param sKey String Field Name
	 * @return Field value or <b>null</b>.
	 * @throws ClassNotFoundException 
	 * @throws SecurityException 
	 * @throws NoSuchMethodException 
	 * @throws InvocationTargetException 
	 * @throws IllegalArgumentException 
	 * @throws IllegalAccessException 
	 * @throws InstantiationException 
	 */
	public static Map<String, String> getMap(JavaRecord oRec, String sKey)
			throws ClassCastException, InstantiationException, IllegalAccessException, IllegalArgumentException,
			InvocationTargetException, NoSuchMethodException, SecurityException, ClassNotFoundException {
		if (oRec.isNull(sKey)) {
			return null;
		} else {
			Object oObj = oRec.get(sKey);
			if (oObj instanceof Map)
				return (Map<String, String>) oObj;
			else
				return (Map<String, String>) Class.forName("org.judal.jdbc.HStore").getConstructor(String.class).newInstance(oObj.toString());
		}
	} // getMap

}

File: scala-adaptor/src/main/scala/org/judal/storage/scala/MapFieldHelper.scala
package org.judal.storage.scala

import java.lang.Iterable

import scala.collection.mutable.Map
import scala.collection.mutable.HashMap

object MapFieldHelper {
 
	/**
	 * <p>Get value of an HStore field<p>
	 * This method is only supported for PostgreSQL HStore fields
	 */
	def getMap(rec: ScalaRecord, key: String) : Map[String,String] = {
	  if (rec.isNull(key)) {
	    null
	  } else {
	    val vmp = new HashMap[String,String]
			val obj = rec.apply(key)
			val cls = Class.forName("org.judal.jdbc.HStore")
			val hst = cls.getConstructor(classOf[String]).newInstance(rec.toString())
			val itr = hst.asInstanceOf[Iterable[java.util.Map.Entry[String, String]]].iterator
			while (itr.hasNext) {
			  val ent = itr.next()
			  vmp.put(ent.getKey, ent.getValue)
			}
	    vmp
	  }
	}
  
}